src/main/java/org/jvnet/hudson/test/LoggerRule.java

```diff
@@ -0,0 +1,129 @@
+import java.util.logging.SimpleFormatter;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.rules.ExternalResource;
+import org.junit.rules.RuleChain;
+
+/**
+ * A test rule which allows you to easily enable one or more loggers for the duration of a test.
+ * Call {@link #record(Class, Level)} or another overload for the rule to take effect.
+ * <p>To capture messages during Jenkins startup,
+ * you may compose this with a {@link JenkinsRule} using a {@link RuleChain};
+ * or use as a {@link ClassRule}.
+ */
+public class LoggerRule extends ExternalResource {
+
+    private final RingBufferLogHandler ringHandler = new RingBufferLogHandler();
+    private final ConsoleHandler consoleHandler = new ConsoleHandler();
+    private final Map<Logger,Level> loggers = new HashMap<Logger,Level>();
+
+    /**
+     * Initializes the rule, by default not recording anything.
+     */
+    public LoggerRule() {
+        ringHandler.setLevel(Level.ALL);
+        consoleHandler.setLevel(Level.ALL);
+    }
+
+    /**
+     * Start listening to a logger.
+     * Might be called in a {@link Rule} initializer, to apply to all test cases in a suite;
+     * or only at the start of selected test cases.
+     * @param logger some logger
+     * @param level something between {@link Level#CONFIG} and {@link Level#ALL};
+     *              using {@link Level#INFO} or above is typically senseless,
+     *              since Java will by default log everything at such levels anyway;
+     *              unless you wish to inspect visible {@link #getRecords},
+     *              or wish to <em>suppress</em> console log output for some logger
+     * @return this rule, for convenience
+     */
+    public LoggerRule record(Logger logger, Level level) {
+        loggers.put(logger, logger.getLevel());
+        logger.setLevel(level);
+        logger.addHandler(ringHandler);
+        logger.addHandler(consoleHandler);
+        return this;
+    }
+
+    /**
+     * Same as {@link #record(Logger, Level) but calls {@link Logger#getLogger(String)} for you first.
+     */
+    public LoggerRule record(String name, Level level) {
+        return record(Logger.getLogger(name), level);
+    }
+    
+    /**
+     * Same as {@link #record(String, Level) but calls {@link Class#getName()} for you first.
+     */
+    public LoggerRule record(Class<?> clazz, Level level) {
+        return record(clazz.getName(), level);
+    }
+
+    /**
+     * Obtains all log records collected so far during this test case.
+     */
+    public List<LogRecord> getRecords() {
+        return ringHandler.getView();
+    }
+
+    /**
+     * Like {@link #getRecords} but applies {@link SimpleFormatter#format(LogRecord)}.
+     */
+    public List<String> getRecordsFormatted() {
+        Formatter f = new SimpleFormatter();
+        List<String> messages = new ArrayList<String>();
+        for (LogRecord lr : getRecords()) {
+            messages.add(f.format(lr).trim());
```

Comment:
For the one demonstrated use case (`StartupTest`), trailing newlines are unwanted. You can always call `getRecords()` if you want to examine precise details. This was more a convenience method to be able to, say, assert that the list of records was empty, or included some substring, etc.